### PR TITLE
My attempt

### PR DESCRIPTION
**Pull request by runkmc**

### CODE REVIEW

**Comment by elubin** [comment]
please include the output file in the checkin

**Comment by elubin** [comment]
having so many private methods, IMO, means you should have a lot more tests.  your only public method hits every one of the private ones, and since you can't test them individually, i think you need to try and figure out ways to test them.

**Comment by mrhead** [comment]
You should not test your private methods! http://www.youtube.com/watch?v=URSWYvyc42M https://gist.github.com/jamesgary/5491390

**Comment by elubin** [comment]
yeah, yeah,.. .i know you don't test private methods.  i meant you do so little in initialize and download_file, that you need to pass in more edge cases to those methods so you exercise all of the private methods for error conditions.

File: bin/analyze.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env ruby
+
+require_relative '../lib/shakespeare_analyzer'
+require_relative '../lib/downloader'
+
+downloader = Downloader.new file: "http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml"
+shakes = ShakespeareAnalyzer.new downloader: downloader
```
Comment (andyw8) [inline bin/analyze.rb]: It might have been nice for the ShakespeareAnalyzer initializer to just accept a string, to avoid it being coupled to the`Downloader` class.

File: lib/downloader.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+require 'open-uri'
+
+class Downloader
+
+  def initialize options = {}
+    @file_to_download = options[:file]
```
Comment (andyw8) [inline lib/downloader.rb]: How about using `fetch` rather than `[]` for accessing the hash? Have a look at [this post](http://devblog.avdi.org/2009/03/16/go-fetch/) by Avdi Grimm if you're not familiar with it.

File: lib/shakespeare_analyzer.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+require 'rexml/document'
```
Comment (andyw8) [inline lib/shakespeare_analyzer.rb:1]: It would be interesting to know why you chose this library over Nokogiri. Are there any trade-offs?

File: lib/shakespeare_analyzer.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  def print_lines
+    in_descending_order(read_lines).each do |key, value|
+      puts "#{key}: #{value} lines"
+    end
+  end
+
+  private
+
+  def get_file
+    xml = @downloader.download_file
```
Comment (andyw8) [inline lib/shakespeare_analyzer.rb]: The `xml` local variable doesn't seem to be needed.

File: lib/shakespeare_analyzer.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  def read_lines 
+    characters = Hash.new 0
```
Comment (andyw8) [inline lib/shakespeare_analyzer.rb:23]: Good call on using Hash.new with the default value.

File: lib/shakespeare_analyzer.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    end
+    characters
+  end
+
+  def current_speaker line
+    line.parent.elements[1].text
+  end
+
+  def in_descending_order characters
+    characters.sort_by { |key, value| -value }
```
Comment (andyw8) [inline lib/shakespeare_analyzer.rb:35]: Nice idea to use `-value` for sorting in descending order. (I used `reverse` after the block).

File: lib/shakespeare_analyzer.rb (hunk 4 above)
Comment (runkmc) [inline lib/shakespeare_analyzer.rb]: Good catch. Thanks!

File: lib/shakespeare_analyzer.rb (hunk 3 above)
Comment (runkmc) [inline lib/shakespeare_analyzer.rb:1]: I don't really know enough about Nokogiri or REXML to know the tradeoffs involved, but since this was a learning exercise I wanted to try to stick to the standard library. That's all.

File: bin/analyze.rb (hunk 1 above)
Comment (runkmc) [inline bin/analyze.rb]: I initially wanted ShakespeareAnalyzer to be concerned only with the xml. I think a good compromise in the meantime would be to make these names a little clearer. "Downloader" is too loaded a term for what happens.

File: lib/downloader.rb (hunk 2 above)
Comment (runkmc) [inline lib/downloader.rb]: Nice.

File: bin/analyze.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env ruby
+
+require_relative '../lib/shakespeare_analyzer'
+require_relative '../lib/downloader'
+
+downloader = Downloader.new file: "http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml"
+shakes = ShakespeareAnalyzer.new xml_provider: downloader
```
Comment (r00k) [inline bin/analyze.rb:7]: I like `xml_provider` as a name!

File: lib/downloader.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+require 'open-uri'
```
Comment (r00k) [inline lib/downloader.rb:1]: Good job isolating this responsibility into one class!

File: lib/shakespeare_analyzer.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
+require 'rexml/document'
+
+class ShakespeareAnalyzer
+
+  def initialize options = {}
+    @xml_provider = options[:xml_provider]
```
Comment (r00k) [inline lib/shakespeare_analyzer.rb:6]: I wouldn't use an options hash for just one argument.

File: lib/shakespeare_analyzer.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+      characters[current_speaker(line)] += 1
+    end
+    characters
+  end
+
+  def current_speaker line
+    line.parent.elements[1].text
+  end
+
+  def in_descending_order characters
```
Comment (r00k) [inline lib/shakespeare_analyzer.rb:34]: Nicely done.

**Comment by r00k** [comment]
Excellent work!

Great job isolating responsibilities at the class level (with Downloader), and at the method level (things like in_descending_order).
